File: source/python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 0:
            return True

        if len(s) % 2 != 0:
            return False

        stack: list[str] = []
        for c in s:
            if c == "(" or c == "[" or c == "{":
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False

                if c == ")":
                    if stack.pop() != "(":
                        return False
                elif c == "]":
                    if stack.pop() != "[":
                        return False
                elif c == "}":
                    if stack.pop() != "{":
                        return False

        return len(stack) == 0


if __name__ == '__main__':
    s = Solution()

    assert s.isValid("()") == True
    assert s.isValid("()[]{}") == True
    assert s.isValid("(]") == False
    assert s.isValid("([)]") == False
    assert s.isValid("{[]}") == True

    print("All tests passed!")

File: source/cpp/20. Valid Parentheses.cpp
#include <iostream>
#include <stack>
#include <string_view>


class Solution {
public:
    bool isValid(std::string_view s) {
        if (s.empty()) return true;
        if (s.size() % 2 != 0) return false;

        std::stack<char> stack;
        for (const auto& c : s) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.empty()) return false;
                const auto top = stack.top();
                stack.pop();
                if (c == ')' && top != '(') return false;
                if (c == ']' && top != '[') return false;
                if (c == '}' && top != '{') return false;
            }
        }

        return stack.empty();
    }
};
